xds/internal/balancer/ringhash/ring.go

```diff
@@ -0,0 +1,174 @@
+// - {idx:9 hash:18151002094784932496 a}
+//
+// To pick from a ring, a binary search will be done for the given target hash,
+// and first item with hash >= given hash will be returned.
+func newRing(subConns map[resolver.Address]*subConn, minRingSize, maxRingSize uint64) (*ring, error) {
+	// https://github.com/envoyproxy/envoy/blob/765c970f06a4c962961a0e03a467e165b276d50f/source/common/upstream/ring_hash_lb.cc#L114
+	normalizedWeights, minWeight, err := normalizeWeights(subConns)
+	if err != nil {
+		return nil, err
+	}
+	// Normalized weightes for {3,3,4} is {0.3,0.3,0.4}.
+
+	// Scale up the size of the ring such that the least-weighted host gets a
+	// whole number of hashes on the ring.
+	//
+	// Note that size is limited by the input max/min.
+	scale := math.Min(math.Ceil(minWeight*float64(minRingSize))/minWeight, float64(maxRingSize))
+	ringSize := math.Ceil(scale)
+	items := make([]*ringEntry, 0, int(ringSize))
+
+	// For each entry, scale*weight nodes are generated in the ring.
+	//
+	// Not all of these are whole numbers. E.g. for weights {a:3,b:3,c:4}, if
+	// ring size is 7, scale is 6.66. The numbers of nodes will be
+	// {a,a,b,b,c,c,c}.
+	//
+	// A hash is generated for each item, and later the results will be sorted
+	// based on the hash.
+	var (
+		idx       int
+		targetIdx float64
+	)
+	for _, scw := range normalizedWeights {
+		targetIdx += scale * scw.weight
+		for float64(idx) < targetIdx {
+			h := xxhash.Sum64String(scw.sc.addr + strconv.Itoa(len(items)))
+			items = append(items, &ringEntry{idx: idx, hash: h, sc: scw.sc})
+			idx++
+		}
+	}
+
+	// Sort items based on hash, to prepare for binary search.
+	sort.Slice(items, func(i, j int) bool { return items[i].hash < items[j].hash })
+	for i, ii := range items {
+		ii.idx = i
+	}
+	return &ring{items: items}, nil
+}
+
+// normalizeWeights divides all the weights by the sum, so that the total weight
+// is 1.
+func normalizeWeights(subConns map[resolver.Address]*subConn) (_ []subConnWithWeight, min float64, _ error) {
+	if len(subConns) == 0 {
+		return nil, 0, fmt.Errorf("number of subconns is 0")
+	}
+	var weightSum uint32
+	for a := range subConns {
+		// The address weight was moved from attributes to the Metadata field.
+		// This is necessary (all the attributes need to be stripped) for the
+		// balancer to detect identical {address+weight} combination.
+		weightSum += a.Metadata.(uint32)
+	}
+	if weightSum == 0 {
+		return nil, 0, fmt.Errorf("total weight of all subconns is 0")
+	}
+	weightSumF := float64(weightSum)
+	ret := make([]subConnWithWeight, 0, len(subConns))
+	first := true
+	for a, sc := range subConns {
+		nw := float64(a.Metadata.(uint32)) / weightSumF
+		ret = append(ret, subConnWithWeight{sc: sc, weight: nw})
+		if first {
+			min = nw
+			first = false
+			continue
+		}
+		if nw < min {
+			min = nw
+		}
+	}
+	// Sort the addresses to return consistent results.
+	//
+	// Note: this might not be necessary? But this makes sure the ring is
```

Comment:
? Is this part of the spec?  What are other languages doing if not?

Reply:
There's no spec. The spec is the Envoy implementation. Envoy's addresses are kept in a vector.

I don't Java is doing this explicitly. They use a hashmap.
C's subchannels are also in a vector, but it seems to be the order of the input addresses.

I think the result is correct either way. This is just a bit nicer, because of map's random order.

Reply:
Maybe remove the "?" then..  `Note: this might not be necessary, but this ...`

Reply:
Done